solr/solrj/src/test/org/apache/solr/common/util/ExecutorUtilTest.java

```diff
@@ -108,6 +111,48 @@ public void testExecutorUtilAwaitsTerminationWhenTaskRespectsInterupt() throws E
     }
   }
 
+  @Test
+  public void testCMDCAwareCachedThreadPool() {
+    // 5 threads max, unbounded queue
+    ExecutorService executor =
+        ExecutorUtil.newMDCAwareCachedThreadPool(
+            5, Integer.MAX_VALUE, new NamedThreadFactory("test"));
+
+    AtomicInteger concurrentTasks = new AtomicInteger();
+    AtomicInteger maxConcurrentTasks = new AtomicInteger();
+    int taskCount = random().nextInt(100);
+
+    for (int i = 0; i < taskCount; i++) {
+      String core = "id_" + random().nextLong();
+
+      Callable<Void> task =
+          () -> {
+            // ensure we never have too many concurrent tasks
+            int concurrent = concurrentTasks.incrementAndGet();
+            assertTrue(concurrent <= 5);
+            maxConcurrentTasks.getAndAccumulate(concurrent, Math::max);
+
+            // assert MDC context is copied from the parent thread that submitted the task
+            assertEquals(core, MDC.get("core"));
+
+            // Sleep for a couple of millis before considering the task is done
+            int delay = random().nextInt(10);
+            Thread.sleep(delay);
+            concurrentTasks.decrementAndGet();
+            return null;
+          };
+
+      MDCLoggingContext.setCoreName(core);
+      executor.submit(task);
+    }
+
+    ExecutorUtil.shutdownAndAwaitTermination(executor);
+
+    // assert the pool was actually multithreaded. Since we submitted many tasks,
+    // all the threads should have been started
+    assertEquals(5, maxConcurrentTasks.get());
```

Comment:
It's possible that taskCount will be < 5 and the test will fail here.

Reply:
Theoretically yes. But that's very unlikely as we submit 1000 tasks.
I consider a decent implementation of the thread pool should always start the 5 requested threads.

Reply:
I added a latch in the test, so we don't depend on the scheduler.

Reply:
Your previous code submitted uniformly random 0-100 tasks, it would have failed 5% of the time. Thank you for adding the `5 + ...` and the latch.